qa/tasks/netem.py

```diff
@@ -0,0 +1,252 @@
+            self.remote.run(args=set_ip)
+
+
+    def link_toggle(self):
+      # For toggling packet drop and recovery in regular interval.
+      # If interval is 5s, link is up for 5s and link is down for 5s
+
+      while not self.stop_event.is_set():
+        self.stop_event.wait(timeout=self.interval)
+        # simulate link down
+        try:
+            self.packet_drop()
+            log.info('link down')
+        except SSHException as e:
+            log.debug('Failed to run command')
+
+        self.stop_event.wait(timeout=self.interval)
+        # if qdisc exist,delete it.
+        try:
+            delete_dev(self.remote)
+            log.info('link up')
+        except SSHException as e:
+            log.debug('Failed to run command')
+
+    def begin(self):
+        self.thread = gevent.spawn(self.link_toggle)
+
+    def end(self):
+        self.stop_event.set()
+        self.thread.get()
+
+
+@contextlib.contextmanager
+def task(ctx, config):
+
+    """
+    - netem:
+          clients: [c1.rgw.0]
+          dst_client: [c2.rgw.1]
+          delay: 10ms
+
+    - netem:
+          clients: [c1.rgw.0]
+          dst_client: [c2.rgw.1]
+          delay_range: [10ms, 20ms] # (min, max)
+
+    - netem:
+          clients: [rgw.1, mon.0]
+          dst_client: [c2.rgw.1]
+          link_toggle_interval: 10 # no unit mentioned. By default takes seconds.
+
+    - netem:
+          clients: [rgw.1, mon.0]
+          link_recover: true
+
+
+    """
+
+    log.info('config %s' % config)
+
+    assert isinstance(config, dict), \
+        "please list clients to run on"
+    if not hasattr(ctx, 'netem'):
+        ctx.netem = argparse.Namespace()
+
+    if config.get('dst_client') is not None:
+        dst = config.get('dst_client')
+        (host,) = ctx.cluster.only(dst).remotes.iterkeys()
+        ctx.netem.host = host
+
+    for role in config.get('clients', None):
+        (remote,) =  ctx.cluster.only(role).remotes.iterkeys()
+        ctx.netem.remote = remote
+        if config.get('delay', False):
+            static_delay(remote, host, config.get('delay'))
+        if config.get('delay_range', False):
+            variable_delay(remote, host, config.get('delay_range'))
+        if config.get('link_toggle_interval', False):
+            log.info('Toggling link for %s' % config.get('link_toggle_interval'))
+            global toggle
+            toggle = Toggle(remote, host, config.get('link_toggle_interval'))
+            ctx.netem.toggle = toggle
+            toggle.begin()
+        if config.get('link_recover', False):
+            log.info('Recovering link')
+            toggle.end()
+            log.info('sleeping')
+            time.sleep(config.get('link_toggle_interval'))
+            delete_dev(ctx.netem.remote)
+
+
+    try:
+        yield
+    finally:
+        if config.get('link_toggle_interval') and not config.get('link_recover'):
+            ctx.netem.toggle.end()
```

Comment:
`link_toggle_interval` and `link_recover` are generally going to be in separate tasks, right? `config` is local to a single task, so you won't see both here. the first task with `link_toggle_interval` will call `toggle.end()` on unwind even if another task with `link_recover` already did

i feel like the task with `link_recover` needs a way to refer to a specific instance of the `netem` task - whether that's by looking for an exact match of `clients`/`dst_client`, or by adding some extra identifier to the yaml format. either way, we'd to store some mapping under `ctx.netem` to locate the `Toggle` for that specific instance

how about a yaml format like this?
```yaml
  netem:
    id: a
    link_toggle_interval: 10
  netem:
    id: b
    delay: 10ms

  ... (other workload tasks)

  netem:
    # undo any toggles or delays configured by netem tasks with id a or b
    link_recover: [a, b]
```

if the first task stored its toggle in `ctx.netem.toggles['a']`, the final task could look it up by id and call `end()` on the correct one

does that make sense?

Reply:
I am still working on this. The goal of link_recover would be to stop the toggle only when link_toggle_interval is set. Wondering if adding an identifier to a task that calls just 'delay' or delay_range' would be necessary since it is going to be cleaned up anyway while unwinding netem.

Reply:
right, i think you'd only need an identifier if you wanted to stop it early with `link_recover`. that may or may not be useful for the `delay` tasks

Reply:
The new patch https://github.com/ceph/ceph/pull/23602/commits/526b705ac5592bc697eb00c3124c3b7d3025566f addresses this.